scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py

```diff
@@ -197,26 +179,31 @@ async def execute_notebook(path: Path, args: argparse.Namespace, config: Config)
     print(f"✅ No problems in {path} (written)")
     return True
 
-async def _execute_notebook(filepath: Path, args: argparse.Namespace) -> nbformat.NotebookNode:
+async def _execute_notebook(filepath: Path, args: argparse.Namespace, patch: bool) -> nbformat.NotebookNode:
     """
-    Use nbconvert to execute notebook.
+    Use nbclient to execute notebook.
     """
     nb = nbformat.read(filepath, as_version=4)
 
-    processor = nbconvert.preprocessors.ExecutePreprocessor(
+    kernel_manager, kernel = await start_new_async_kernel(
+        kernel_name="python3",
+        extra_arguments=["--InlineBackend.figure_format='svg'"],
+    )
+
+    if patch:
+        kernel.execute(MOCKING_CODE, store_history=False)
+
+    notebook_client = nbclient.NotebookClient(
+        nb=nb,
+        km=kernel_manager,
+        kc=kernel,
         # If submitting jobs, we want to wait forever (-1 means no timeout)
         timeout=-1 if args.submit_jobs else 300,
-        kernel_name="python3",
-        extra_arguments=["--InlineBackend.figure_format='svg'"]
     )
 
     # This runs the notebook, including possibly submitting jobs. We run it in a
     # new thread to avoid blocking other notebooks from submitting jobs.
-    with patch_runtime(nb, should_patch=not args.submit_jobs):
-        await asyncio.to_thread(processor.preprocess, nb)
-
-    if not args.write:
-        return nb
+    await notebook_client.async_execute()
```

Comment:
I think we need to reset the kernel between each notebook, otherwise we share state? This is `reset_kc` https://nbclient.readthedocs.io/en/latest/reference/nbclient.html#nbclient.client.NotebookClient.async_execute

Also the comment above about using a thread is no longer accurate. Imo you don't need a comment at all - it's already obvious enough with async/await that we're getting concurrency thanks to its usage. Before, we were explicitly wrapping synchronous code with an async context.

Reply:
This function creates a new kernel and runs the notebook in it, so each notebook will get its own kernel. I'll update the docstring to make that clearer.

Reply:
Ahh, yeah I read that poorly. Agreed. Thanks!

Reply:
If you missed it then someone else will probably too, so best to make it clearer